File: android/app/src/main/java/app/releaf/mobile/features/contacts/ContactsViewModel.kt
/*
 * ContactsViewModel.kt
 *
 * Backs the contacts tab. Two data sources:
 *   - `ContactDirectoryRepository` — aggregated app contacts
 *     (notepad entries + pages). Always-on observer.
 *   - `DeviceContactsProvider` — device address book, queried
 *     on-demand as the user types. Gated by `READ_CONTACTS`.
 *
 * Exposes a single [ContactsUiState] with both halves so the
 * screen renders without having to coordinate two flows itself.
 */

package app.releaf.mobile.features.contacts

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import app.releaf.mobile.ReleafApp
import app.releaf.mobile.auth.AuthStore
import app.releaf.mobile.data.contact.ContactDirectoryRepository
import app.releaf.mobile.data.contact.DeviceContactsProvider
import app.releaf.mobile.data.contact.DirectoryContact
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

data class ContactsUiState(
    val query: String = "",
    val isLoading: Boolean = true,
    /** Full app contacts (no filter). */
    val allAppContacts: List<DirectoryContact> = emptyList(),
    /** App contacts filtered by [query]. When query is blank, equals [allAppContacts]. */
    val filteredAppContacts: List<DirectoryContact> = emptyList(),
    /** Device contacts matching [query]. Empty when no query, no permission, or no matches. */
    val deviceContacts: List<DirectoryContact> = emptyList(),
    /** True when `READ_CONTACTS` has been granted. */
    val devicePermissionGranted: Boolean = false,
) {
    val isSearching: Boolean get() = query.trim().isNotEmpty()
    val totalAppContacts: Int get() = allAppContacts.size
}

@OptIn(ExperimentalCoroutinesApi::class, FlowPreview::class)
class ContactsViewModel(
    application: Application,
    private val userIdProvider: () -> String?,
    private val directoryRepository: ContactDirectoryRepository,
    private val deviceContactsProvider: DeviceContactsProvider,
) : AndroidViewModel(application) {

    private val _query = MutableStateFlow("")
    val query: StateFlow<String> = _query.asStateFlow()

    private val _permissionGranted = MutableStateFlow(deviceContactsProvider.hasPermission())
    private val _deviceContacts = MutableStateFlow<List<DirectoryContact>>(emptyList())

    private val appContactsFlow = run {
        val userId = userIdProvider()
        if (userId.isNullOrBlank()) flowOf(emptyList())
        else directoryRepository.observeAll(userId)
    }

    val state: StateFlow<ContactsUiState> = combine(
        _query,
        appContactsFlow,
        _deviceContacts,
        _permissionGranted,
    ) { q, appContacts, deviceHits, granted ->
        val trimmed = q.trim()
        val filtered = if (trimmed.isEmpty()) {
            appContacts
        } else {
            // Rank contacts whose *name* starts with the query
            // above contacts that merely contain it elsewhere. The
            // secondary key keeps each rank bucket alphabetical.
            appContacts
                .filter { it.matches(trimmed) }
                .sortedByMatchRank(trimmed)
        }
        ContactsUiState(
            query                    = q,
            isLoading                = false,
            allAppContacts           = appContacts,
            filteredAppContacts      = filtered,
            deviceContacts           = deviceHits,
            devicePermissionGranted  = granted,
        )
    }.stateIn(
        scope        = viewModelScope,
        started      = SharingStarted.WhileSubscribed(5_000),
        initialValue = ContactsUiState(),
    )

    private var deviceSearchJob: Job? = null

    fun updateQuery(value: String) {
        _query.value = value
        scheduleDeviceSearch(value)
    }

    fun clearQuery() {
        _query.value = ""
        deviceSearchJob?.cancel()
        _deviceContacts.value = emptyList()
    }

    /** Called after the permission request resolves. */
    fun onPermissionResult(granted: Boolean) {
        _permissionGranted.value = granted
        // Re-run the search so previously-empty device results
        // populate instantly after the user grants access.
        scheduleDeviceSearch(_query.value)
    }

    private fun scheduleDeviceSearch(rawQuery: String) {
        deviceSearchJob?.cancel()
        val q = rawQuery.trim()
        if (q.isEmpty() || !deviceContactsProvider.hasPermission()) {
            _deviceContacts.value = emptyList()
            return
        }
        deviceSearchJob = viewModelScope.launch {
            kotlinx.coroutines.delay(150)
            // Device search returns a LIKE-style hit list; reorder
            // so names starting with the query appear before
            // mid-string matches, mirroring the app-contact
            // behaviour above.
            _deviceContacts.value = deviceContactsProvider.search(q).sortedByMatchRank(q)
        }
    }

    companion object {
        val Factory: ViewModelProvider.Factory = viewModelFactory {
            initializer {
                val app = this[ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY] as ReleafApp
                ContactsViewModel(
                    application             = app,
                    userIdProvider          = { (app.authStore.state.value as? app.releaf.mobile.auth.AuthState.SignedIn)?.session?.userId },
                    directoryRepository     = app.contactDirectoryRepository,
                    deviceContactsProvider  = app.deviceContactsProvider,
                )
            }
        }
    }
}

private fun DirectoryContact.matches(query: String): Boolean {
    return name.contains(query, ignoreCase = true) ||
        phones.any { it.contains(query, ignoreCase = true) } ||
        email?.contains(query, ignoreCase = true) == true ||
        organization?.contains(query, ignoreCase = true) == true
}

/**
 * Rank matching contacts so that names starting with the query
 * surface before names that only contain the query mid-string.
 * Each rank bucket is sorted alphabetically on the name.
 */
internal fun List<DirectoryContact>.sortedByMatchRank(
    query: String,
): List<DirectoryContact> {
    val lc = query.lowercase()
    return sortedWith(
        compareBy<DirectoryContact> { !it.name.lowercase().startsWith(lc) }
            .thenBy(String.CASE_INSENSITIVE_ORDER) { it.name }
    )
}

File: apps/releaf/ios/Releaf/Features/Contacts/ContactsViewModel.swift
/*
 * ContactsViewModel.swift
 *
 * Backs the Contacts screen. Observes the aggregated app directory
 * and, on user input, searches the device address book. Coalesces
 * both halves into a single `ContactsState` the view renders.
 */

import Foundation
import ReleafData

@MainActor
public final class ContactsViewModel: ObservableObject {

    public struct State: Equatable {
        public var query: String = ""
        public var isLoading: Bool = true
        public var allAppContacts: [DirectoryContact] = []
        public var filteredAppContacts: [DirectoryContact] = []
        public var deviceContacts: [DirectoryContact] = []
        public var devicePermissionGranted: Bool = false

        public var isSearching: Bool {
            !query.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }
    }

    @Published public private(set) var state: State = State()

    private let userId: String
    private let directory: ContactDirectoryRepository
    private let device: DeviceContactsProvider

    private var directoryTask: Task<Void, Never>?
    private var deviceSearchTask: Task<Void, Never>?

    public init(
        userId: String,
        directory: ContactDirectoryRepository = ContactDirectoryRepository(),
        device: DeviceContactsProvider = DeviceContactsProvider()
    ) {
        self.userId = userId
        self.directory = directory
        self.device = device
        self.state.devicePermissionGranted = device.hasPermission
    }

    public func start() {
        stop()
        directoryTask = Task { [weak self, directory, userId] in
            guard let self else { return }
            do {
                for try await all in directory.observeAll(userId: userId) {
                    self.state.isLoading = false
                    self.state.allAppContacts = all
                    self.recomputeFiltered()
                }
            } catch {}
        }
    }

    public func stop() {
        directoryTask?.cancel()
        directoryTask = nil
        deviceSearchTask?.cancel()
        deviceSearchTask = nil
    }

    deinit {
        directoryTask?.cancel()
        deviceSearchTask?.cancel()
    }

    public func updateQuery(_ value: String) {
        state.query = value
        recomputeFiltered()
        scheduleDeviceSearch()
    }

    public func clearQuery() {
        state.query = ""
        state.deviceContacts = []
        recomputeFiltered()
        deviceSearchTask?.cancel()
    }

    public func requestPermission() {
        Task { [weak self, device] in
            let granted = await device.requestPermission()
            await MainActor.run {
                self?.state.devicePermissionGranted = granted
                self?.scheduleDeviceSearch()
            }
        }
    }

    // MARK: - Private

    private func recomputeFiltered() {
        let trimmed = state.query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            state.filteredAppContacts = state.allAppContacts
            return
        }
        let lc = trimmed.lowercased()
        let matches = state.allAppContacts.filter { c in
            if c.name.lowercased().contains(lc) { return true }
            if c.phones.contains(where: { $0.lowercased().contains(lc) }) { return true }
            if let email = c.email, email.lowercased().contains(lc) { return true }
            if let org = c.organization, org.lowercased().contains(lc) { return true }
            return false
        }
        // Names that *start* with the query rank above names that
        // only contain it mid-string. Within each bucket the list
        // stays alphabetical.
        state.filteredAppContacts = matches.sortedByMatchRank(query: lc)
    }

    private func scheduleDeviceSearch() {
        deviceSearchTask?.cancel()
        let q = state.query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !q.isEmpty, device.hasPermission else {
            state.deviceContacts = []
            return
        }
        deviceSearchTask = Task { [weak self, device] in
            try? await Task.sleep(nanoseconds: 150_000_000)
            if Task.isCancelled { return }
            let hits = await device.search(rawQuery: q)
            let ranked = hits.sortedByMatchRank(query: q.lowercased())
            await MainActor.run {
                self?.state.deviceContacts = ranked
            }
        }
    }
}

// MARK: - Rank helper

internal extension Array where Element == DirectoryContact {
    /// Rank matching contacts so names that *start* with [query]
    /// surface before names that only contain it mid-string.
    /// Each rank bucket is sorted alphabetically on the name.
    func sortedByMatchRank(query lc: String) -> [DirectoryContact] {
        sorted { lhs, rhs in
            let lStarts = lhs.name.lowercased().hasPrefix(lc)
            let rStarts = rhs.name.lowercased().hasPrefix(lc)
            if lStarts != rStarts { return lStarts }
            return lhs.name.localizedCaseInsensitiveCompare(rhs.name) == .orderedAscending
        }
    }
}
